src/components/TrackListItem.js

```diff
@@ -23,77 +23,81 @@ export const TrackListItem = ({
-    function reducer(state, action) {
-      let newState = {...state};
-      switch (action.type){
-        case "setType":
-          newState["trackFile"] = undefined;
-          newState["trackType"] = action.payload;
-          break;
-        case "setFile":
-          newState["trackFile"] = action.payload;
-          break;
-        case "setColorSettings":
-          newState["trackColorSettings"] = action.payload;
-          break;
-        default:
-          throw new Error();
-      }
-      return newState;
-    }
   
     const trackTypeOnChange = async(newTrackType) => {
       // wait until file change to call onChange
-      dispatch({type: "setType", payload: newTrackType});
+      let newPropChanges = {...propChanges};
+      newPropChanges["trackType"] = newTrackType;
+      newPropChanges["trackFile"] = undefined;
+      setPropChanges(newPropChanges);
     }
 
     const trackFileOnChange = async(newFile) => {
-      dispatch({type: "setFile", payload: newFile});
+      let newPropChanges = {...propChanges};
+      newPropChanges["trackFile"] = newFile;
+      setPropChanges(newPropChanges);
     }
 
     const trackSettingsOnChange = async(key, value) => {
       let newTrackColorSettings = {...trackProps["trackColorSettings"]};
       newTrackColorSettings[key] = value;
-      dispatch({type: "setColorSettings", payload: newTrackColorSettings});
+      let newPropChanges = {...propChanges};
+      newPropChanges["trackColorSettings"] = newTrackColorSettings;
+      setPropChanges(newPropChanges);
     }
 
-    // https://stackoverflow.com/questions/55840294/how-to-fix-missing-dependency-warning-when-using-useeffect-react-hook
     // useEffect hook to tell react to call onchange after state changes
+    // setPropChanges are async, useEffect acts as a callback to setPropChanges
+    // https://stackoverflow.com/questions/54954091/how-to-use-callback-with-usestate-hook-in-react/56394177#56394177
     useEffect(() => {
-      if (myTrackProps.trackFile !== undefined && JSON.stringify(trackProps) !== JSON.stringify(myTrackProps)) {
-        _onChange.current(myTrackProps);
+      // get an updated version of trackProps
+      let newTrackProps = {...trackProps};
+      for (const key in propChanges) {
+        newTrackProps[key] = propChanges[key];
       }
       
-    }, [myTrackProps.trackFile, myTrackProps.trackType, myTrackProps.trackColorSettings, myTrackProps, _onChange, trackProps]);
+      // push it if trackFile has been selected and changes are made
+      if (newTrackProps.trackFile !== undefined && JSON.stringify(trackProps) !== JSON.stringify(newTrackProps)) {
+        onChange(trackID, newTrackProps);
+        setPropChanges({});
+      }
+    }, [propChanges, onChange, trackProps, trackID]);
 
+    // displayed elements uses propChanges(local state) first, then uses trackProps
+    const displayedFile = "trackFile" in propChanges ? propChanges["trackFile"] : trackProps["trackFile"];
```

Comment:
This isn't just an `||` because `propChanges` is supposed to be able to make the file undefined, right?